app/controllers/templates_controller.rb

```diff
@@ -79,15 +79,21 @@ def auto_complete_controller_name
   end
 
   def preview
-    base = @template.preview_host_collection
-    @host = params[:preview_host_id].present? ? base.find(params[:preview_host_id]) : base.first
+    if params[:id]
+      find_resource
+      base = @template.preview_host_collection
+      @host = params[:preview_host_id].present? ? base.find(params[:preview_host_id]) : base.first
+    else
+      type = request.original_url.split("/")[4]
+      @host = Host.find(params[:preview_host_id])
+      @template = resource_class.new(params[type])
+    end
     if @host.nil?
       render :text => _('No host could be found for rendering the template'), :status => :not_found
       return
     end
-
     @template.template = params[:template]
-    safe_render(@template)
+    safe_render(params[:id] ? @template : @template.delete)
```

Comment:
I'm a bit confused by this line. Correct me if I'm wrong, this is how I understand it:

```ruby
`@template` at this point is either defined by `find_resource` or above `resource_class.new(params[type])`
If `params[:id].present?` it renders `@template`. That's the usual behavior and it's fine.
If `params[:id].nil?` it renders `@template.delete`. In this case, `@template` isn't saved yet on the database, as it was created from `resource_class.new`, why does it have to be deleted?
``` 
Wouldn't `safe_render(@template)` still work in both cases here? The template created in line 89 for hosts isn't ever saved as far as I can tell.